signature/src/balrog.rs

```diff
@@ -0,0 +1,363 @@
+        if list.is_empty() {
+            return Err(BalrogError::CertificateNotFound);
+        }
+
+        let mut parsed: Vec<X509Certificate<'a>> = Vec::new();
+        let mut hash: Vec<u8> = Vec::new();
+        for (i, pem) in list.iter().enumerate() {
+            if pem.label != "CERTIFICATE" {
+                return Err(BalrogError::from(X509Error::InvalidCertificate));
+            }
+
+            let x509 = pem.parse_x509()?;
+            parsed.push(x509);
+
+            /* For the final certificate, calculate its SHA256 hash. */
+            /* TODO: Technically, we can get more bytes in the PEM beyond the end
+             * of the ASN.1 certificate sequence, and this function will include
+             * them in the hash. This feels like an unclear edge case though.
+             */
+            if i + 1 == list.len() {
+                let digest = digest::digest(&digest::SHA256, pem.contents.as_slice());
+                hash.extend_from_slice(digest.as_ref());
+            }
+        }
+
+        Ok(Balrog {
+            chain: parsed,
+            root_hash: hash,
+        })
+    }
+
+    /* Ensure that the certificate chain is valid. */
+    pub fn verify_chain(&self, current_time: i64, root_hash: &str) -> Result<(), BalrogError> {
+        if self.chain.is_empty() {
+            return Err(BalrogError::CertificateNotFound);
+        }
+
+        /* For each certificate N in the chain, check that:
+         *  1. The issuer name in cert N matches the subject name of cert N+1.
+         *  2. The validity time of cert N contains the current time.
+         *  3. The cert N+1 contains a key usage extension permitting CA.
+         *  4. Extract the public key from cert N+1
+         *  5. Call X509Certificate::verify_signature with the key from step 4.
+         *
+         * For the last certificate in the chain (the root), repeat the same
+         * steps using the last certificate as both cert N and cert N+1, and
+         * check that the SHA256 of the root certificate matches the root_hash
+         * parameter to this function.
+         */
+        let mut index = 0;
+        while index + 1 < self.chain.len() {
+            let subject = &self.chain[index];
+            let issuer = &self.chain[index + 1];
+            let _ = Self::verify_cert_chain_pair(subject, issuer, current_time)?;
+            index = index + 1;
+        }
+
+        /* Verify the root certificate. */
+        let root_cert = match self.chain.last() {
+            None => return Err(BalrogError::CertificateNotFound),
+            Some(x) => x,
+        };
+        let _ = Self::verify_cert_chain_pair(root_cert, root_cert, current_time)?;
+
+        let hash_stripped = root_hash.replace(":", "");
+        let hash_decoded = match hex::decode(hash_stripped) {
+            Err(_e) => return Err(BalrogError::RootHashParseFailed),
+            Ok(x) => x,
+        };
+        if self.root_hash.is_empty() || (self.root_hash != hash_decoded) {
```

Comment:
nit: Should we just check if the root_hash is empty at the top of the function?

Reply:
I think it's fine, there are fewer return statements to follow doing it this way.